### PR TITLE
Remove unnecessary uses of irep_id_hash

### PR DESCRIPTION
**Pull request by owen-mc-diffblue**
#2082 means that we don't need to specify a hash function to `irep_idt`. This PR removes unnecessary uses of `irep_id_hash`. It also builds on #2075 and removes typedefs for `std::set<irep_idt>` and `std::unordered_set<irep_idt>` on the basis that they are now short enough to use directly.

### CODE REVIEW

**Comment by tautschnig** [comment]
Could you please add "Fixes: #2090" to one of your commit messages?

File: src/ansi-c/expr2c_class.h [hunk 1]
```diff
@@ -64,9 +64,8 @@ class expr2ct
 
   static std::string indent_str(unsigned indent);
 
-  std::unordered_map<irep_idt,
-                std::unordered_set<irep_idt, irep_id_hash>,
-                irep_id_hash> ns_collision;
+  std::unordered_map<irep_idt, std::unordered_set<irep_idt>, irep_id_hash>
```
Comment (tautschnig) [inline src/ansi-c/expr2c_class.h]: I don't think you need the `irep_id_hash` here anymore.

File: src/goto-instrument/code_contracts.cpp [hunk 2]
```diff
@@ -389,8 +388,8 @@ void code_contractst::operator()()
     goto_functions.function_map.find(CPROVER_PREFIX "initialize");
   assert(i_it!=goto_functions.function_map.end());
 
-  for(id_sett::const_iterator it=summarized.begin();
-      it!=summarized.end();
+  for(std::unordered_set<irep_idt>::const_iterator it = summarized.begin();
```
Comment (tautschnig) [inline src/goto-instrument/code_contracts.cpp]: Possibly in a separate commit, but this should likely be changed to `for(const auto &contract : summarized)`

File: src/goto-instrument/dump_c.cpp [hunk 3]
```diff
@@ -769,9 +769,8 @@ void dump_ct::dump_typedefs(std::ostream &os) const
   // output
   std::map<std::string, typedef_infot> to_insert;
 
-  typedef std::unordered_set<irep_idt, irep_id_hash> id_sett;
-  id_sett typedefs_done;
-  std::unordered_map<irep_idt, id_sett, irep_id_hash>
+  std::unordered_set<irep_idt> typedefs_done;
+  std::unordered_map<irep_idt, std::unordered_set<irep_idt>, irep_id_hash>
```
Comment (tautschnig) [inline src/goto-instrument/dump_c.cpp]: Drop `irep_id_hash`

File: src/goto-programs/remove_unused_functions.cpp [hunk 4]
```diff
@@ -57,8 +57,7 @@ void find_used_functions(
   goto_functionst &functions,
   std::set<irep_idt> &seen)
 {
-  std::pair<std::set<irep_idt>::const_iterator, bool> res =
-    seen.insert(start);
+  std::pair<std::set<irep_idt>::const_iterator, bool> res = seen.insert(start);
```
Comment (tautschnig) [inline src/goto-programs/remove_unused_functions.cpp]: Nit pick: Looks like an unrelated whitespace change.

File: src/linking/linking.cpp [hunk 5]
```diff
@@ -1186,9 +1187,9 @@ void linkingt::do_type_dependencies(id_sett &needs_to_be_renamed)
 
   std::stack<irep_idt> queue;
 
-  for(id_sett::const_iterator
-      d_it=needs_to_be_renamed.begin();
-      d_it!=needs_to_be_renamed.end();
+  for(std::unordered_set<irep_idt>::const_iterator d_it =
+        needs_to_be_renamed.begin();
+      d_it != needs_to_be_renamed.end();
```
Comment (tautschnig) [inline src/linking/linking.cpp]: Possibly in a separate commit: `std::stack<irep_idt> queue(needs_to_be_renamed.begin(), needs_to_be_renamed.end());` and ditch the `for` loop altogether.

File: src/linking/linking.cpp [hunk 6]
```diff
@@ -1197,11 +1198,10 @@ void linkingt::do_type_dependencies(id_sett &needs_to_be_renamed)
...
     queue.pop();
 
-    const id_sett &u=used_by[id];
+    const std::unordered_set<irep_idt> &u = used_by[id];
 
-    for(id_sett::const_iterator
-        d_it=u.begin();
-        d_it!=u.end();
+    for(std::unordered_set<irep_idt>::const_iterator d_it = u.begin();
+        d_it != u.end();
```
Comment (tautschnig) [inline src/linking/linking.cpp]: Possibly separate commit: `for(const auto &dep : u)`

File: src/linking/linking_class.h [hunk 7]
```diff
@@ -173,12 +171,15 @@ class linkingt:public typecheckt
   namespacet ns;
 
   // X -> Y iff Y uses X for new symbol type IDs X and Y
-  typedef std::unordered_map<irep_idt, id_sett, irep_id_hash> used_byt;
+  typedef std::unordered_map<irep_idt,
+                             std::unordered_set<irep_idt>,
+                             irep_id_hash>
```
Comment (tautschnig) [inline src/linking/linking_class.h]: Drop `irep_id_hash`

**Review by tautschnig — APPROVED** [review]
Looks good to me! Could you confirm that doing `git grep id_sett` doesn't yield anything that's left?

File: src/linking/linking.cpp (hunk 5 above)
Comment (owen-mc-diffblue) [reply src/linking/linking.cpp]: It seems that `std::stack` doesn't have a constructor like that. The solution seems to be to use `std::deque` instead.

**Comment by owen-mc-diffblue** [comment]
The output of `git grep id_sett` is non-empty because there are some typedefs of `id_sett` to something else, and a few typedefs to an identifier containing "id_sett" as a substring. Here is the output of `git grep id_sett | grep typedef`
```
src/cpp/cpp_scope.h:  typedef std::set<cpp_idt *> id_sett;
src/cpp/cpp_scopes.h:  typedef std::set<cpp_idt *> id_sett;
src/cpp/expr2cpp.cpp:  typedef std::unordered_set<std::string, string_hash> id_sett;
src/goto-instrument/object_id.h:typedef std::set<object_idt> object_id_sett;
src/goto-instrument/points_to.h:  typedef std::map<object_idt, object_id_sett> value_mapt;
src/pointer-analysis/value_set.h:  typedef std::set<unsigned int> dynamic_object_id_sett;
src/pointer-analysis/value_set_fi.h:  typedef std::unordered_set<unsigned int> dynamic_object_id_sett;
src/pointer-analysis/value_set_fivr.h:  typedef std::unordered_set<unsigned int> dynamic_object_id_sett;
src/pointer-analysis/value_set_fivrns.h:  typedef std::unordered_set<unsigned int> dynamic_object_id_sett;
```

**Comment by owen-mc-diffblue** [comment]
I may as well post the whole output of `git grep id_sett`:
```
src/cpp/cpp_declarator_converter.cpp:      cpp_scopet::id_sett id_set;
src/cpp/cpp_declarator_converter.cpp:    cpp_scopest::id_sett id_set;
src/cpp/cpp_declarator_converter.cpp:    for(cpp_scopest::id_sett::const_iterator
src/cpp/cpp_instantiate_template.cpp:    cpp_scopet::id_sett id_set;
src/cpp/cpp_scope.cpp:  id_sett &id_set)
src/cpp/cpp_scope.cpp:  id_sett &id_set)
src/cpp/cpp_scope.cpp:  id_sett &id_set)
src/cpp/cpp_scope.cpp:  id_sett id_set;
src/cpp/cpp_scope.h:  typedef std::set<cpp_idt *> id_sett;
src/cpp/cpp_scope.h:    id_sett &id_set);
src/cpp/cpp_scope.h:    id_sett &id_set);
src/cpp/cpp_scope.h:    id_sett &id_set);
src/cpp/cpp_scopes.h:  typedef std::set<cpp_idt *> id_sett;
src/cpp/cpp_typecheck_compound_type.cpp:  cpp_scopet::id_sett id_set;
src/cpp/cpp_typecheck_compound_type.cpp:    cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_resolve.cpp:  const cpp_scopest::id_sett &id_set,
src/cpp/cpp_typecheck_resolve.cpp:  for(cpp_scopest::id_sett::const_iterator
src/cpp/cpp_typecheck_resolve.cpp:      cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_resolve.cpp:  const cpp_scopest::id_sett &id_set,
src/cpp/cpp_typecheck_resolve.cpp:  for(cpp_scopest::id_sett::const_iterator
src/cpp/cpp_typecheck_resolve.cpp:  for(cpp_scopest::id_sett::const_iterator
src/cpp/cpp_typecheck_resolve.cpp:  cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_resolve.cpp:  cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_resolve.cpp:    for(cpp_scopest::id_sett::const_iterator
src/cpp/cpp_typecheck_resolve.cpp:      cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_resolve.cpp:      for(cpp_scopest::id_sett::const_iterator it=id_set.begin();
src/cpp/cpp_typecheck_resolve.cpp:        cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_resolve.cpp:        for(cpp_scopest::id_sett::const_iterator
src/cpp/cpp_typecheck_resolve.cpp:  cpp_scopest::id_sett &id_set)
src/cpp/cpp_typecheck_resolve.cpp:  cpp_scopest::id_sett new_set;
src/cpp/cpp_typecheck_resolve.cpp:  for(cpp_scopest::id_sett::const_iterator
src/cpp/cpp_typecheck_resolve.cpp:  cpp_scopest::id_sett &id_set)
src/cpp/cpp_typecheck_resolve.cpp:  for(cpp_scopest::id_sett::iterator
src/cpp/cpp_typecheck_resolve.cpp:      cpp_scopest::id_sett::iterator old(it);
src/cpp/cpp_typecheck_resolve.cpp:  cpp_scopest::id_sett &id_set,
src/cpp/cpp_typecheck_resolve.cpp:    cpp_scopest::id_sett tmp_set;
src/cpp/cpp_typecheck_resolve.h:    const cpp_scopest::id_sett &id_set,
src/cpp/cpp_typecheck_resolve.h:    const cpp_scopest::id_sett &id_set,
src/cpp/cpp_typecheck_resolve.h:    cpp_scopest::id_sett &id_set,
src/cpp/cpp_typecheck_resolve.h:  void filter_for_named_scopes(cpp_scopest::id_sett &id_set);
src/cpp/cpp_typecheck_resolve.h:  void filter_for_namespaces(cpp_scopest::id_sett &id_set);
src/cpp/cpp_typecheck_template.cpp:    cpp_scopet::id_sett id_set;
src/cpp/cpp_typecheck_template.cpp:  cpp_scopet::id_sett id_set;
src/cpp/cpp_typecheck_template.cpp:  cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_template.cpp:  for(cpp_scopest::id_sett::iterator
src/cpp/cpp_typecheck_template.cpp:    cpp_scopest::id_sett::iterator next=it;
src/cpp/cpp_typecheck_template.cpp:    cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_using.cpp:  cpp_scopest::id_sett id_set;
src/cpp/cpp_typecheck_using.cpp:  for(cpp_scopest::id_sett::iterator
src/cpp/expr2cpp.cpp:  typedef std::unordered_set<std::string, string_hash> id_sett;
src/goto-instrument/object_id.cpp:  object_id_sett &dest,
src/goto-instrument/object_id.cpp:void get_objects(const exprt &expr, object_id_sett &dest)
src/goto-instrument/object_id.cpp:void get_objects_r(const code_assignt &assign, object_id_sett &dest)
src/goto-instrument/object_id.cpp:void get_objects_w(const code_assignt &assign, object_id_sett &dest)
src/goto-instrument/object_id.cpp:void get_objects_w(const exprt &lhs, object_id_sett &dest)
src/goto-instrument/object_id.cpp:void get_objects_r_lhs(const exprt &lhs, object_id_sett &dest)
src/goto-instrument/object_id.h:typedef std::set<object_idt> object_id_sett;
src/goto-instrument/object_id.h:void get_objects(const exprt &expr, object_id_sett &dest);
src/goto-instrument/object_id.h:void get_objects_r(const code_assignt &assign, object_id_sett &);
src/goto-instrument/object_id.h:void get_objects_w(const code_assignt &assign, object_id_sett &);
src/goto-instrument/object_id.h:void get_objects_w_lhs(const exprt &lhs, object_id_sett &);
src/goto-instrument/object_id.h:void get_objects_r_lhs(const exprt &lhs, object_id_sett &);
src/goto-instrument/points_to.cpp:    for(object_id_sett::const_iterator
src/goto-instrument/points_to.h:  const object_id_sett &operator[](const object_idt &object_id)
src/goto-instrument/points_to.h:  typedef std::map<object_idt, object_id_sett> value_mapt;
src/goto-instrument/points_to.h:  const object_id_sett empty_set;
src/pointer-analysis/value_set.cpp:  dynamic_object_id_sett to_mark;
src/pointer-analysis/value_set.h:  typedef std::set<unsigned int> dynamic_object_id_sett;
src/pointer-analysis/value_set_fi.cpp:  dynamic_object_id_sett to_mark;
src/pointer-analysis/value_set_fi.h:  typedef std::unordered_set<unsigned int> dynamic_object_id_sett;
src/pointer-analysis/value_set_fivr.cpp:  dynamic_object_id_sett to_mark;
src/pointer-analysis/value_set_fivr.h:  typedef std::unordered_set<unsigned int> dynamic_object_id_sett;
src/pointer-analysis/value_set_fivrns.cpp:  dynamic_object_id_sett to_mark;
src/pointer-analysis/value_set_fivrns.h:  typedef std::unordered_set<unsigned int> dynamic_object_id_sett;
```

**Review by kroening — APPROVED** [review]

**Comment by tautschnig** [comment]
Thank you very much for the follow-up check! The only one that strikes me as problematic is `src/cpp/expr2cpp.cpp:  typedef std::unordered_set<std::string, string_hash> id_sett;` - could you please remove this as it simply is unused?

**Review by tautschnig — APPROVED** [review]
